### PR TITLE
Make tests use NewEnvelope instead of NewEnvelopeConn.

### PR DESCRIPTION
**Pull request by ghemawat**
Introduce a new Child interface that NewEnvelope uses to run the weavelet. Typically this is a real process, but tests that want to run the weavelet in-process can supply a custom implementation.

### CODE REVIEW

File: runtime/envelope/child.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	Start(context.Context, *protos.AppConfig) error
+
+	// Wait for the child to exit.
+	// REQUIRES: Start has been called.
+	// REQUIRES: Wait has not been called.
+	Wait() error
+
+	// Different IO streams connecting us to the child
+	Reader() io.ReadCloser  // Can be used to receive messages from the Child
+	Writer() io.WriteCloser // Can be used to receive messages from the Child
```
Comment (mwhittaker) [inline runtime/envelope/child.go:43]:
```suggestion
	Writer() io.WriteCloser // Can be used to send messages to the Child
```

File: runtime/envelope/envelope.go [hunk 2]
```diff
@@ -153,19 +155,24 @@ func NewEnvelope(ctx context.Context, wlet *protos.EnvelopeInfo, config *protos.
...
-	}
-	runtime.OnExitSignal(func() { os.RemoveAll(tmpDir) }) // Cleanup when process exits
-
-	// Arrange to delete tmpDir if this function returns an error.
-	removeDir := true // Cleared on a successful return
-	defer func() {
-		if removeDir {
-			os.RemoveAll(tmpDir)
+	var removeDir bool
+	tmpDir := options.TmpDir
```
Comment (mwhittaker) [inline runtime/envelope/envelope.go:159]: If a user provides the `TmpDir` option, are we promising not to delete the directory? It seems like here, we don't delete it, but on line 257 in the `Serve` method, we do?

File: internal/tool/multi/deployer.go [hunk 3]
```diff
@@ -353,8 +353,10 @@ func (d *deployer) startColocationGroup(g *group) error {
 			d.stop(err)
 			return err
 		})
-		if err := d.registerReplica(g, wlet, e.Pid()); err != nil {
-			return err
+		if pid, ok := e.Pid(); ok {
```
Comment (mwhittaker) [inline internal/tool/multi/deployer.go:356]: What do you think about panicking if `!ok`? If I understand correctly, the multi deployer should never create an `Envelope` with an in-process `Child`, so `ok` should always be true. And if `ok` isn't true and we don't call `registerReplica`, I think the deployer may end up in an unexpected state and either crash or hang. 

And same for the SSH deployer.

**Review by mwhittaker — APPROVED** [review]
Very nice! I like the `Child` API and am happy to see a lot of duplicated `EnvelopeConn` code go away :partying_face:

File: runtime/envelope/child.go (hunk 1 above)
Comment (ghemawat) [reply runtime/envelope/child.go]: Done

File: runtime/envelope/envelope.go (hunk 2 above)
Comment (ghemawat) [reply runtime/envelope/envelope.go:160]: Great catch. Thanks.

File: internal/tool/multi/deployer.go (hunk 3 above)
Comment (ghemawat) [reply internal/tool/multi/deployer.go]: Done

**Review by ghemawat — COMMENTED** [review]
Thanks!

**Pull request merged**